File: src/LCR044.hPov7L.1000297/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function largestValues (root: TreeNode | null): number[] {
    if (!root) {
        return [];
    }
    const result:number[] = [];
    let nodes = [root, ];
    while (nodes.length) {
        const nNodes:TreeNode[] = [];
        let max = -Infinity;
        for (const node of nodes) {
            max = Math.max(max, node.val);
            if (node.left) {
                nNodes.push(node.left);
            }
            if (node.right) {
                nNodes.push(node.right);
            }
        }
        nodes = nNodes;
        result.push(max);
    }
    return result;
}

File: src/LCR044.hPov7L.1000297/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private final List<Integer> result = new ArrayList<>();
    public List<Integer> largestValues(TreeNode root) {
        if(root == null){
            return result;
        }
        levelOrder(List.of(root));
        return result;
    }
    
    private void levelOrder(List<TreeNode> nodes){
        List<TreeNode> next = new ArrayList<>();
        int max = Integer.MIN_VALUE;
        for(TreeNode node : nodes){
            max = Math.max(max,node.val);
            if(node.left != null){
                next.add(node.left);
            }
            if(node.right != null){
                next.add(node.right);
            }
        }
        result.add(max);
        
        if(!next.isEmpty()){
            levelOrder(next);
        }
    }
}